src/main/java/eu/europeana/metis/sandbox/service/dataset/DatasetReportServiceImpl.java

```diff
@@ -41,197 +43,205 @@
+                errorsLog);
+
+        // collect steps processing information
+        List<ProgressByStepDto> stepsInfo = new LinkedList<>();
+        recordsProcessedByStep.forEach((step, statusMap) -> addStepInfo(stepsInfo, statusMap, step,
+                recordErrorsByStep));
+
+        return new ProgressInfoDto(
+                getPublishPortalUrl(dataset, completedRecords),
+                dataset.getRecordsQuantity(), completedRecords,
+                stepsInfo, datasetInfoDto, getErrorMessage(completedRecords, failedRecords, dataset));
     }
 
-    // get qty of records completely processed
-    long completedRecords = getCompletedRecords(stepStatistics);
-
-    // get qty of records that failed
-    long failedRecords = getFailedRecords(stepStatistics);
-
-    // get records processed by step
-    Map<Step, Map<Status, Long>> recordsProcessedByStep = getStatisticsByStep(
-        stepStatistics);
-
-    // get errors by step
-    Map<Step, Map<Status, Map<String, List<ErrorLogView>>>> recordErrorsByStep = getRecordErrorsByStep(
-        errorsLog);
-
-    // collect steps processing information
-    List<ProgressByStepDto> stepsInfo = new LinkedList<>();
-    recordsProcessedByStep.forEach((step, statusMap) -> addStepInfo(stepsInfo, statusMap, step,
-        recordErrorsByStep));
-
-    return new ProgressInfoDto(
-        getPublishPortalUrl(dataset, completedRecords, failedRecords),
-        dataset.getRecordsQuantity(), completedRecords,
-        stepsInfo, datasetInfoDto);
-  }
-
-  private String getPublishPortalUrl(DatasetEntity dataset, Long completedRecords,
-      Long failedRecords) {
-    return getPortalUrl(portalPublishDatasetUrl, dataset, completedRecords, failedRecords);
-  }
-
-  private String getPortalUrl(String portal, DatasetEntity dataset, Long completedRecords,
-      Long failedRecords) {
-    Long recordsQty = dataset.getRecordsQuantity();
-    if (recordsQty == null) {
-      return HARVESTING_IDENTIFIERS_MESSAGE;
+    private String getPublishPortalUrl(DatasetEntity dataset, Long completedRecords) {
+        return getPortalUrl(portalPublishDatasetUrl, dataset, completedRecords);
     }
-    if (recordsQty == 0) {
-      return EMPTY_DATASET_MESSAGE;
+
+    private String getPortalUrl(String portal, DatasetEntity dataset, Long completedRecords) {
+        Long recordsQty = dataset.getRecordsQuantity();
+        if (recordsQty == null) {
+            return HARVESTING_IDENTIFIERS_MESSAGE;
+        }
+
+        if (!recordsQty.equals(completedRecords)) {
+            return PROCESSING_DATASET_MESSAGE;
+        }
+
+        var datasetId = dataset.getDatasetId() + SEPARATOR + dataset.getDatasetName() + SUFFIX;
+        return portal + URLEncoder.encode(datasetId, StandardCharsets.UTF_8);
     }
-    if (!recordsQty.equals(completedRecords)) {
-      return PROCESSING_DATASET_MESSAGE;
+
+    private String getErrorMessage(Long completedRecords, Long failedRecords, DatasetEntity dataset) {
```

Comment:
my suggestion is: ```private String getErrorMessage(Long completedRecords, Long failedRecords, Long recordsQty) ``` and you can remove the line ```Long recordsQty = dataset.getRecordsQuantity();```

Reply:
Done